### PR TITLE
change exchange trade history

### PR DESCRIPTION
**Pull request by halink0803**
Move include exchanges into reserve data instead of put it in http object.

### CODE REVIEW

**Comment by halink0803** [comment]
[Change api history design](https://trello.com/c/BCqSgwjM/40-change-api-history-design)

File: data/reserve_data.go [hunk 1]
```diff
@@ -306,16 +307,35 @@ func (self ReserveData) ControlAuthDataSize() error {
...
+func (self ReserveData) GetTradeHistory(fromTime, toTime uint64) (common.AllTradeHistory, error) {
+	var err error
+	data := common.AllTradeHistory{
+		Timestamp: common.GetTimestamp(),
+		Data:      map[common.ExchangeID]common.ExchangeTradeHistory{},
+	}
+	for _, ex := range self.exchanges {
+		history, err := ex.GetTradeHistory(fromTime, toTime)
+		if err != nil {
+			return data, err
```
Comment (favadi) [inline data/reserve_data.go]: should we return nil, err? Because we should never evaluate data if err != nil.

File: data/reserve_data.go [hunk 2]
```diff
@@ -306,16 +307,35 @@ func (self ReserveData) ControlAuthDataSize() error {
...
 }
 
+func (self ReserveData) GetTradeHistory(fromTime, toTime uint64) (common.AllTradeHistory, error) {
+	var err error
+	data := common.AllTradeHistory{
+		Timestamp: common.GetTimestamp(),
+		Data:      map[common.ExchangeID]common.ExchangeTradeHistory{},
+	}
+	for _, ex := range self.exchanges {
+		history, err := ex.GetTradeHistory(fromTime, toTime)
```
Comment (favadi) [inline data/reserve_data.go]: this err instance will shadow the err instance in L311. Please remove L311 as you don't assign it to anything in this function.

File: http/server.go [hunk 3]
```diff
@@ -846,17 +845,9 @@ func (self *HTTPServer) GetTradeHistory(c *gin.Context) {
...
-	for _, ex := range self.exchanges {
-		history, err := ex.GetTradeHistory(fromTime, toTime)
-		if err != nil {
-			httputil.ResponseFailure(c, httputil.WithError(err))
-			return
-		}
-		data.Data[ex.ID()] = history
+	data, err := self.app.GetTradeHistory(fromTime, toTime)
+	if err != nil {
+		httputil.ResponseFailure(c, httputil.WithReason(err.Error()))
```
Comment (favadi) [inline http/server.go]: you added the httputil.WithError option right? Please use it. :)

File: data/reserve_data.go (hunk 1 above)
Comment (halink0803) [reply data/reserve_data.go]: We have define return AllTradeHistory. Cannot return nil, plus we also check err before evaluate data then it should have encountered error here.

File: data/reserve_data.go (hunk 2 above)
Comment (halink0803) [reply data/reserve_data.go]: Yep, done.

File: http/server.go (hunk 3 above)
Comment (halink0803) [reply http/server.go]: Yep, I used it. Actually I did not add WithError function.

File: data/reserve_data.go (hunk 1 above)
Comment (favadi) [reply data/reserve_data.go]: @halink0803 In L311, I see that data is initialized with non zero value.

Comment (halink0803) [reply data/reserve_data.go]: Saw it. Change it to zero value.

**Pull request merged**